source/extensions/filters/udp/dns_filter/dns_parser.h

```diff
@@ -0,0 +1,182 @@
+
+  virtual ~BaseDnsRecord() = default;
+  void serializeName(Buffer::OwnedImpl& output);
+  virtual void serialize(Buffer::OwnedImpl& output) PURE;
+
+  const std::string name_;
+  const uint16_t type_;
+  const uint16_t class_;
+};
+
+/**
+ * DnsQueryRecord represents a query record parsed from a DNS request from a client. Each record
+ * contains the ID, domain requested and the flags dictating the type of record that is sought.
+ */
+class DnsQueryRecord : public BaseDnsRecord {
+public:
+  DnsQueryRecord(const std::string& rec_name, const uint16_t rec_type, const uint16_t rec_class)
+      : BaseDnsRecord(rec_name, rec_type, rec_class) {}
+  void serialize(Buffer::OwnedImpl& output) override;
+};
+
+using DnsQueryRecordPtr = std::unique_ptr<DnsQueryRecord>;
+using DnsQueryPtrVec = std::vector<DnsQueryRecordPtr>;
+
+using AddressConstPtrVec = std::vector<Network::Address::InstanceConstSharedPtr>;
+using AnswerCallback = std::function<void(DnsQueryRecordPtr& query, AddressConstPtrVec& ipaddr)>;
+
+/**
+ * DnsAnswerRecord represents a single answer record for a name that is to be serialized and sent to
+ * a client. This class differs from the BaseDnsRecord and DnsQueryRecord because it contains
+ * additional fields for the TTL and address.
+ */
+class DnsAnswerRecord : public BaseDnsRecord {
+public:
+  DnsAnswerRecord(const std::string& query_name, const uint16_t rec_type, const uint16_t rec_class,
+                  const uint32_t ttl, Network::Address::InstanceConstSharedPtr ipaddr)
+      : BaseDnsRecord(query_name, rec_type, rec_class), ttl_(ttl), ip_addr_(ipaddr) {}
+  void serialize(Buffer::OwnedImpl& output) override { UNREFERENCED_PARAMETER(output); }
+
+  const uint32_t ttl_;
+  const Network::Address::InstanceConstSharedPtr ip_addr_;
+};
+
+using DnsAnswerRecordPtr = std::unique_ptr<DnsAnswerRecord>;
+using DnsAnswerMap = std::unordered_multimap<std::string, DnsAnswerRecordPtr>;
+
+/**
+ * DnsQueryContext contains all the data associated with a query. The filter uses this object to
+ * generate a response and determine where it should be transmitted.
+ */
+class DnsQueryContext {
+public:
+  DnsQueryContext(Network::Address::InstanceConstSharedPtr local,
+                  Network::Address::InstanceConstSharedPtr peer)
+      : local_(std::move(local)), peer_(std::move(peer)), parse_status_(false), id_() {}
+  ~DnsQueryContext() = default;
+
+  const Network::Address::InstanceConstSharedPtr local_;
+  const Network::Address::InstanceConstSharedPtr peer_;
+
+  bool parse_status_;
+  uint16_t id_;
+  DnsQueryPtrVec queries_;
+  DnsAnswerMap answers_;
+};
+
+using DnsQueryContextPtr = std::unique_ptr<DnsQueryContext>;
+
+/**
+ * This class orchestrates parsing a DNS query and building the response to be sent to a client.
+ */
+class DnsMessageParser : public Logger::Loggable<Logger::Id::filter> {
+public:
+  enum DnsResponseCode { NoError, FormatError, ServerFailure, NameError, NotImplemented };
```

Comment:
enum class, same elsewhere if I missed any of these.

Reply:
I will change these to constexpr.   I used enum w/o class since I'm using the numeric values of the entries.